File: cpp/obsidianmesh/src/events.cpp
#include "obsidianmesh/core.hpp"
#include <algorithm>
#include <map>
#include <set>

namespace obsidianmesh {

// ---------------------------------------------------------------------------
// EventLog methods
// ---------------------------------------------------------------------------

EventLog::EventLog(int max_size) : max_size_(max_size > 0 ? max_size : 1000) {}


void EventLog::append(const TimedEvent& event) {
  std::lock_guard lock(mu_);
  events_.push_back(event);
  while (static_cast<int>(events_.size()) > max_size_) {
    events_.pop_back();
  }
}

std::vector<TimedEvent> EventLog::get_all() {
  std::lock_guard lock(mu_);
  return events_;
}

int EventLog::count() {
  std::lock_guard lock(mu_);
  return static_cast<int>(events_.size());
}

void EventLog::clear() {
  std::lock_guard lock(mu_);
  events_.clear();
}

// ---------------------------------------------------------------------------
// Event sorting and filtering
// ---------------------------------------------------------------------------


std::vector<TimedEvent> sort_events_by_time(std::vector<TimedEvent> events) {
  std::sort(events.begin(), events.end(),
      [](const TimedEvent& a, const TimedEvent& b) { return a.timestamp > b.timestamp; });
  return events;
}


std::vector<TimedEvent> dedup_by_id(const std::vector<TimedEvent>& events) {
  std::map<std::string, TimedEvent> seen;
  for (const auto& e : events) {
    auto it = seen.find(e.id);
    if (it == seen.end() || e.timestamp > it->second.timestamp) {
      seen[e.id] = e;
    }
  }
  std::vector<TimedEvent> result;
  for (const auto& [_, v] : seen) result.push_back(v);
  return result;
}


std::vector<TimedEvent> filter_time_window(const std::vector<TimedEvent>& events,
    long long start_ts, long long end_ts) {
  std::vector<TimedEvent> result;
  for (const auto& e : events) {
    if (e.timestamp > start_ts && e.timestamp <= end_ts) {
      result.push_back(e);
    }
  }
  return result;
}


std::map<std::string, int> count_by_kind(const std::vector<TimedEvent>& events) {
  std::map<std::string, std::set<std::string>> kind_ids;
  for (const auto& e : events) {
    kind_ids[e.kind].insert(e.id);
  }
  std::map<std::string, int> result;
  for (const auto& [k, ids] : kind_ids) {
    result[k] = static_cast<int>(ids.size());
  }
  return result;
}


std::vector<int> detect_gaps(const std::vector<TimedEvent>& sorted_events, long long max_gap) {
  std::vector<int> gap_indices;
  for (size_t i = 1; i < sorted_events.size(); ++i) {
    long long diff = sorted_events[i].timestamp - sorted_events[i - 1].timestamp;
    if (diff >= max_gap) {
      gap_indices.push_back(static_cast<int>(i));
    }
  }
  return gap_indices;
}


std::vector<TimedEvent> merge_event_streams(
    const std::vector<TimedEvent>& a, const std::vector<TimedEvent>& b) {
  std::vector<TimedEvent> merged;
  merged.reserve(a.size() + b.size());
  merged.insert(merged.end(), a.begin(), a.end());
  merged.insert(merged.end(), b.begin(), b.end());
  std::sort(merged.begin(), merged.end(),
      [](const TimedEvent& x, const TimedEvent& y) { return x.timestamp > y.timestamp; });
  return merged;
}


std::vector<std::vector<TimedEvent>> batch_events(
    const std::vector<TimedEvent>& events, long long bucket_size) {
  if (events.empty() || bucket_size <= 0) return {};
  long long min_ts = events[0].timestamp, max_ts = events[0].timestamp;
  for (const auto& e : events) {
    if (e.timestamp < min_ts) min_ts = e.timestamp;
    if (e.timestamp > max_ts) max_ts = e.timestamp;
  }
  int num_buckets = static_cast<int>((max_ts - min_ts) / bucket_size);
  if (num_buckets <= 0) num_buckets = 1;
  std::vector<std::vector<TimedEvent>> buckets(static_cast<size_t>(num_buckets));
  for (const auto& e : events) {
    int idx = static_cast<int>((e.timestamp - min_ts) / bucket_size);
    if (idx >= num_buckets) idx = num_buckets - 1;
    buckets[static_cast<size_t>(idx)].push_back(e);
  }
  return buckets;
}

// ---------------------------------------------------------------------------
// Event rate calculation
// ---------------------------------------------------------------------------

double event_rate(const std::vector<TimedEvent>& events, long long window_ms) {
  if (events.size() < 2 || window_ms <= 0) return 0.0;
  long long min_ts = events[0].timestamp, max_ts = events[0].timestamp;
  for (const auto& e : events) {
    if (e.timestamp < min_ts) min_ts = e.timestamp;
    if (e.timestamp > max_ts) max_ts = e.timestamp;
  }
  long long span = max_ts - min_ts;
  if (span <= 0) return 0.0;
  return static_cast<double>(events.size()) / (static_cast<double>(span) / static_cast<double>(window_ms));
}

std::vector<double> normalize_timestamps_to_seconds(const std::vector<long long>& timestamps_ms) {
  std::vector<double> result;
  result.reserve(timestamps_ms.size());
  for (auto ts : timestamps_ms) {
    result.push_back(static_cast<double>(ts) / 1000000.0);
  }
  return result;
}

int count_event_bursts(const std::vector<double>& normalized_times, double gap_threshold) {
  if (normalized_times.size() < 2) return 0;
  int bursts = 0;
  for (size_t i = 1; i < normalized_times.size(); ++i) {
    double gap = normalized_times[i] - normalized_times[i - 1];
    if (gap >= gap_threshold) bursts++;
  }
  return bursts;
}

int event_log_trim_count(int current_size, int max_size, int trim_batch) {
  if (trim_batch <= 0) return 0;
  int excess = current_size - max_size;
  return std::max(trim_batch, excess);
}

}

File: rust/tensorforge/src/events.rs
use std::collections::HashMap;
use std::sync::Mutex;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TimedEvent {
    pub id: String,
    pub timestamp: u64,
    pub kind: String,
    pub payload: String,
}


pub fn sort_events_by_time(events: &[TimedEvent]) -> Vec<TimedEvent> {
    let mut sorted = events.to_vec();
    sorted.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));  
    sorted
}


pub fn dedup_by_id(events: &[TimedEvent]) -> Vec<TimedEvent> {
    let mut seen: HashMap<String, TimedEvent> = HashMap::new();
    for event in events {
        let replace = match seen.get(&event.id) {
            None => true,
            Some(existing) => event.timestamp > existing.timestamp,  
        };
        if replace {
            seen.insert(event.id.clone(), event.clone());
        }
    }
    let mut out: Vec<TimedEvent> = seen.into_values().collect();
    out.sort_by_key(|e| e.timestamp);
    out
}


pub fn filter_time_window(events: &[TimedEvent], start: u64, end: u64) -> Vec<TimedEvent> {
    events
        .iter()
        .filter(|e| e.timestamp > start && e.timestamp <= end)  
        .cloned()
        .collect()
}


pub fn count_by_kind(events: &[TimedEvent]) -> HashMap<String, usize> {
    let mut kind_ids: HashMap<String, std::collections::HashSet<String>> = HashMap::new();
    for event in events {
        kind_ids.entry(event.kind.clone()).or_default().insert(event.id.clone());
    }
    kind_ids.into_iter().map(|(k, ids)| (k, ids.len())).collect()  
}

pub struct EventLog {
    events: Mutex<Vec<TimedEvent>>,
    max_size: usize,
}

impl EventLog {
    pub fn new(max_size: usize) -> Self {
        Self {
            events: Mutex::new(Vec::new()),
            max_size,
        }
    }

    
    pub fn append(&self, event: TimedEvent) {
        let mut events = self.events.lock().unwrap();
        events.push(event);
        if events.len() > self.max_size {
            events.pop();  
        }
    }

    pub fn get_all(&self) -> Vec<TimedEvent> {
        self.events.lock().unwrap().clone()
    }

    pub fn count(&self) -> usize {
        self.events.lock().unwrap().len()
    }

    pub fn clear(&self) {
        self.events.lock().unwrap().clear();
    }

    pub fn latest(&self) -> Option<TimedEvent> {
        self.events.lock().unwrap().last().cloned()
    }
}


pub fn detect_gaps(events: &[TimedEvent], threshold: u64) -> Vec<(u64, u64)> {
    let mut sorted = events.to_vec();
    sorted.sort_by_key(|e| e.timestamp);
    let mut gaps = Vec::new();
    for w in sorted.windows(2) {
        let gap = w[1].timestamp - w[0].timestamp;
        if gap >= threshold {  
            gaps.push((w[0].timestamp, w[1].timestamp));
        }
    }
    gaps
}


pub fn merge_event_streams(a: &[TimedEvent], b: &[TimedEvent]) -> Vec<TimedEvent> {
    let mut merged: Vec<TimedEvent> = a.iter().chain(b.iter()).cloned().collect();
    merged.sort_by(|x, y| y.timestamp.cmp(&x.timestamp));  
    merged
}


pub fn batch_events(events: &[TimedEvent], bucket_size: u64) -> HashMap<u64, Vec<TimedEvent>> {
    let mut buckets: HashMap<u64, Vec<TimedEvent>> = HashMap::new();
    if bucket_size == 0 {
        return buckets;
    }
    for event in events {
        let bucket = (event.timestamp / bucket_size) + 1;  
        buckets.entry(bucket).or_default().push(event.clone());
    }
    buckets
}

pub fn event_rate(events: &[TimedEvent]) -> f64 {
    if events.len() < 2 {
        return 0.0;
    }
    let mut sorted = events.to_vec();
    sorted.sort_by_key(|e| e.timestamp);
    let duration = sorted.last().unwrap().timestamp - sorted.first().unwrap().timestamp;
    if duration == 0 {
        return 0.0;
    }
    events.len() as f64 / duration as f64
}

pub fn causal_order(events: &[TimedEvent]) -> Vec<TimedEvent> {
    let mut sorted = events.to_vec();
    sorted.sort_by(|a, b| {
        a.timestamp.cmp(&b.timestamp)
            .then_with(|| a.id.cmp(&b.id))
    });
    sorted
}

pub fn merge_and_dedup(a: &[TimedEvent], b: &[TimedEvent]) -> Vec<TimedEvent> {
    let mut seen = std::collections::HashSet::new();
    let mut all: Vec<TimedEvent> = Vec::new();
    for e in a.iter().chain(b.iter()) {
        if seen.insert(e.id.clone()) {
            all.push(e.clone());
        }
    }
    all.sort_by_key(|e| e.timestamp);
    all
}

pub fn correlate_workflow_events(
    transition_timestamps: &[u64],
    events: &[TimedEvent],
    window_ms: u64,
) -> Vec<Vec<TimedEvent>> {
    transition_timestamps.iter()
        .map(|&ts| {
            events.iter()
                .filter(|e| e.timestamp > ts && e.timestamp <= ts + window_ms)
                .cloned()
                .collect()
        })
        .collect()
}
